paragraph/scheduling/logger.cc

```diff
@@ -0,0 +1,105 @@
+/* Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "paragraph/scheduling/logger.h"
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include "paragraph/scheduling/instruction_fsm.h"
+
+namespace paragraph {
+
+shim::StatusOr<std::unique_ptr<Logger>> Logger::Create(
+    const std::string& filename) {
+  auto logger = absl::WrapUnique(new Logger(filename));
+  if (filename != "") {
+    RETURN_IF_ERROR(logger->OpenFile());
+    RETURN_IF_ERROR(logger->InitializeCsv());
+  }
+  return logger;
+}
+
+void Logger::FlushToFile() {
+  if (filename_ != "") {
+    if (ostream_.is_open()) {
+      ostream_.close();
+    }
+  }
+}
+
+bool Logger::IsAvailable() {
+  if (ostream_.is_open()) {
+    return true;
+  } else if (filename_ != "") {
+    return false;
+  } else if (OpenFile().ok()) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+absl::Status Logger::SetFilename(const std::string& filename) {
+  FlushToFile();
+  filename_ = filename;
+  RETURN_IF_ERROR(OpenFile());
+  RETURN_IF_ERROR(InitializeCsv());
+  return absl::OkStatus();
+}
+
+absl::Status Logger::AppendToCsv(const InstructionFsm& instruction_fsm) {
+  RETURN_IF_ERROR(OpenFile());
+  std::cout << instruction_fsm.ToCsv() << std::endl;
```

Comment:
I don't think you should have the instruction FSM handle the format. This logger class specifies the format in the "InitializeCsv" function and the ToCsv function would need to track that format. Neither could be adjusted independently. I think you should move the ToCsv functionality into the Logger class.